File: scripts/mssql_to_mysql_copy.py
#!/usr/bin/env python3
"""
Read-only MSSQL → MySQL copier for selected tables.

Features
- Connects to SQL Server (READ ONLY usage; only selects).
- Creates destination MySQL tables with compatible types (if missing).
- Copies rows in batches to avoid memory issues.

Requirements
- Python 3.9+
- pip install pyodbc mysql-connector-python
- SQL Server ODBC Driver installed and reachable (for pyodbc)

Usage
  python3 scripts/mssql_to_mysql_copy.py --config scripts/config.example.json
  python3 scripts/mssql_to_mysql_copy.py --config scripts/config.example.json --tables CariEtiketTanimi,StokOzellikleri

Notes
- Ensure the SQL Server user has only SELECT permissions on the source DB.
- This script never issues INSERT/UPDATE/DELETE on the source.
- Destination tables are created only if missing. Set overwrite=false to keep existing.
"""
import argparse
import json
import os
import sys
from typing import Any, Dict, List, Optional, Tuple

try:
    import pyodbc  # type: ignore
except Exception as e:  # pragma: no cover
    print("Missing dependency: pyodbc. Run: pip install pyodbc", file=sys.stderr)
    raise

try:
    import mysql.connector  # type: ignore
except Exception as e:  # pragma: no cover
    print("Missing dependency: mysql-connector-python. Run: pip install mysql-connector-python", file=sys.stderr)
    raise


def debug(msg: str) -> None:
    print(msg)


def load_config(path: str) -> Dict[str, Any]:
    with open(path, "r", encoding="utf-8") as f:
        cfg = json.load(f)
    # Defaults
    cfg.setdefault("source", {})
    cfg.setdefault("target", {})
    cfg.setdefault("options", {})
    cfg.setdefault("tables", [])
    cfg["source"].setdefault("schema", "dbo")
    cfg["target"].setdefault("create_database", False)
    cfg["options"].setdefault("batch_size", 1000)
    cfg["options"].setdefault("overwrite", False)
    cfg["options"].setdefault("engine", "InnoDB")
    cfg["options"].setdefault("charset", "utf8mb4")
    cfg["options"].setdefault("collation", None)
    return cfg


def mssql_connection_string(source_cfg: Dict[str, Any]) -> str:
    # Prefer explicit DRIVER. Example for macOS/Linux with msodbcsql17:
    # DRIVER={ODBC Driver 17 for SQL Server};SERVER=host,1433;DATABASE=GEMAS_DYS;UID=...;PWD=...
    driver = source_cfg.get("driver") or "ODBC Driver 17 for SQL Server"
    server = source_cfg.get("server")
    port = source_cfg.get("port")
    database = source_cfg.get("database")
    uid = source_cfg.get("user")
    pwd = source_cfg.get("password")
    app_intent = source_cfg.get("application_intent", "ReadOnly")

    if not (server and database and uid and pwd):
        raise ValueError("source.server, source.database, source.user, source.password are required")
    server_part = f"{server},{port}" if port else server
    # ApplicationIntent is honored when connecting to Availability Groups; it's still safe to include.
    conn = (
        "DRIVER={{{}}};SERVER={};DATABASE={};UID={};PWD={};ApplicationIntent={};TrustServerCertificate=Yes;".format(
            driver, server_part, database, uid, pwd, app_intent
        )
    )
    return conn


def connect_mssql(source_cfg: Dict[str, Any]) -> pyodbc.Connection:
    conn_str = mssql_connection_string(source_cfg)
    debug("Connecting MSSQL (read-only usage)...")
    cn = pyodbc.connect(conn_str, autocommit=False)
    # Set read-only session if driver supports it (best-effort)
    try:
        SQL_ATTR_ACCESS_MODE = 101  # pyodbc attr
        SQL_MODE_READ_ONLY = 1
        cn.set_attr(SQL_ATTR_ACCESS_MODE, SQL_MODE_READ_ONLY)
    except Exception:
        pass
    return cn


def connect_mysql(target_cfg: Dict[str, Any]) -> mysql.connector.MySQLConnection:
    debug("Connecting MySQL (target)...")
    required = ["host", "database", "user", "password"]
    for k in required:
        if not target_cfg.get(k):
            raise ValueError(f"target.{k} is required")
    conn = mysql.connector.connect(
        host=target_cfg["host"],
        port=target_cfg.get("port", 3306),
        database=target_cfg["database"],
        user=target_cfg["user"],
        password=target_cfg["password"],
        autocommit=False,
        charset=target_cfg.get("charset", "utf8mb4"),
        use_pure=True,
    )
    return conn


def fetch_columns_mssql(cn: pyodbc.Connection, db: str, schema: str, table: str) -> List[Dict[str, Any]]:
    sql = """
    SELECT
      c.COLUMN_NAME,
      c.DATA_TYPE,
      c.CHARACTER_MAXIMUM_LENGTH,
      c.NUMERIC_PRECISION,
      c.NUMERIC_SCALE,
      c.IS_NULLABLE
    FROM INFORMATION_SCHEMA.COLUMNS c
    WHERE c.TABLE_CATALOG = ? AND c.TABLE_SCHEMA = ? AND c.TABLE_NAME = ?
    ORDER BY c.ORDINAL_POSITION
    """
    cur = cn.cursor()
    cur.execute(sql, (db, schema, table))
    rows = cur.fetchall()
    cols = []
    for r in rows:
        cols.append(
            {
                "COLUMN_NAME": r[0],
                "DATA_TYPE": r[1],
                "CHARACTER_MAXIMUM_LENGTH": r[2],
                "NUMERIC_PRECISION": r[3],
                "NUMERIC_SCALE": r[4],
                "IS_NULLABLE": r[5],
            }
        )
    if not cols:
        raise RuntimeError(f"No columns found for {schema}.{table} in {db}")
    return cols


def fetch_pk_mssql(cn: pyodbc.Connection, db: str, schema: str, table: str) -> List[str]:
    sql = """
    SELECT kcu.COLUMN_NAME
    FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS tc
    JOIN INFORMATION_SCHEMA.KEY_COLUMN_USAGE kcu
      ON tc.CONSTRAINT_NAME = kcu.CONSTRAINT_NAME
     AND tc.TABLE_SCHEMA = kcu.TABLE_SCHEMA
     AND tc.TABLE_NAME = kcu.TABLE_NAME
    WHERE tc.TABLE_CATALOG = ?
      AND tc.TABLE_SCHEMA = ?
      AND tc.TABLE_NAME = ?
      AND tc.CONSTRAINT_TYPE = 'PRIMARY KEY'
    ORDER BY kcu.ORDINAL_POSITION
    """
    cur = cn.cursor()
    cur.execute(sql, (db, schema, table))
    return [r[0] for r in cur.fetchall()]


def map_sqlserver_type_to_mysql(col: Dict[str, Any]) -> str:
    dt = str(col["DATA_TYPE"]).lower()
    length = col["CHARACTER_MAXIMUM_LENGTH"]
    precision = col["NUMERIC_PRECISION"]
    scale = col["NUMERIC_SCALE"]

    if dt in ("int",):
        return "INT"
    if dt in ("bigint",):
        return "BIGINT"
    if dt in ("smallint",):
        return "SMALLINT"
    if dt in ("tinyint",):
        return "TINYINT"
    if dt in ("bit",):
        return "TINYINT(1)"
    if dt in ("real",):
        return "FLOAT"
    if dt in ("float",):
        return "DOUBLE"
    if dt in ("decimal", "numeric", "money", "smallmoney"):
        p = precision or 19
        s = scale if scale is not None else 4
        return f"DECIMAL({int(p)},{int(s)})"
    if dt in ("char", "nchar"):
        if length == -1:
            return "LONGTEXT"
        return f"CHAR({int(length)})"
    if dt in ("varchar", "nvarchar"):
        if length == -1:
            return "LONGTEXT"
        return f"VARCHAR({int(length)})"
    if dt in ("text", "ntext"):
        return "LONGTEXT"
    if dt in ("varbinary",):
        if length == -1:
            return "LONGBLOB"
        return f"VARBINARY({int(length)})"
    if dt in ("binary",):
        return f"BINARY({int(length)})" if length and length > 0 else "BINARY(1)"
    if dt in ("image",):
        return "LONGBLOB"
    if dt in ("date",):
        return "DATE"
    if dt in ("smalldatetime", "datetime"):
        return "DATETIME"
    if dt in ("datetime2",):
        return "DATETIME(6)"
    if dt in ("time",):
        return "TIME"
    if dt in ("uniqueidentifier",):
        return "CHAR(36)"
    if dt in ("xml",):
        return "LONGTEXT"
    # Fallback
    return "LONGTEXT"


def mysql_table_exists(cur, table: str) -> bool:
    cur.execute("SHOW TABLES LIKE %s", (table,))
    return cur.fetchone() is not None


def quote_ident_mysql(name: str) -> str:
    return f"`{name.replace('`','``')}`"


def build_create_table_mysql(table: str, columns: List[Dict[str, Any]], pk_cols: List[str], opts: Dict[str, Any]) -> str:
    cols_sql = []
    for col in columns:
        name = col["COLUMN_NAME"]
        null_ok = str(col["IS_NULLABLE"]).upper() == "YES"
        mapped = map_sqlserver_type_to_mysql(col)
        cols_sql.append(f"{quote_ident_mysql(name)} {mapped} {'NULL' if null_ok else 'NOT NULL'}")

    pk_sql = f", PRIMARY KEY ({', '.join(quote_ident_mysql(c) for c in pk_cols)})" if pk_cols else ""
    engine = opts.get("engine", "InnoDB")
    charset = opts.get("charset", "utf8mb4")
    collation = opts.get("collation")
    collate_sql = f" COLLATE={collation}" if collation else ""
    ddl = (
        f"CREATE TABLE {quote_ident_mysql(table)} (\n  "
        + ",\n  ".join(cols_sql)
        + pk_sql
        + f"\n) ENGINE={engine} DEFAULT CHARSET={charset}{collate_sql};"
    )
    return ddl


def copy_table(
    mssql_cn: pyodbc.Connection,
    mysql_cn: mysql.connector.MySQLConnection,
    db: str,
    schema: str,
    table: str,
    options: Dict[str, Any],
) -> None:
    debug(f"\n=== {schema}.{table} ===")
    cols = fetch_columns_mssql(mssql_cn, db, schema, table)
    pk_cols = fetch_pk_mssql(mssql_cn, db, schema, table)

    mycur = mysql_cn.cursor()
    exists = mysql_table_exists(mycur, table)
    if not exists:
        ddl = build_create_table_mysql(table, cols, pk_cols, options)
        debug(f"Creating table {table} in MySQL...")
        mycur.execute(ddl)
        mysql_cn.commit()
    else:
        if options.get("overwrite"):
            debug(f"Overwriting table {table} (dropping + recreating)...")
            mycur.execute(f"DROP TABLE {quote_ident_mysql(table)}")
            mysql_cn.commit()
            ddl = build_create_table_mysql(table, cols, pk_cols, options)
            mycur.execute(ddl)
            mysql_cn.commit()
        else:
            debug(f"Table {table} already exists; appending rows.")

    # Prepare select and insert
    col_names = [c["COLUMN_NAME"] for c in cols]
    src_select = (
        f"SELECT "
        + ", ".join(f"[{n}]" for n in col_names)
        + f" FROM [{schema}].[{table}]"
    )
    placeholders = ", ".join(["%s"] * len(col_names))
    insert_sql = f"INSERT INTO {quote_ident_mysql(table)} (" + ", ".join(quote_ident_mysql(n) for n in col_names) + f") VALUES ({placeholders})"

    batch_size = int(options.get("batch_size", 1000))
    cur = mssql_cn.cursor()
    cur.fast_executemany = False  # select only; not used on mssql cursor
    cur.execute(src_select)

    total = 0
    batch: List[Tuple[Any, ...]] = []
    while True:
        rows = cur.fetchmany(batch_size)
        if not rows:
            break
        for r in rows:
            batch.append(tuple(r))
        mycur.executemany(insert_sql, batch)
        mysql_cn.commit()
        total += len(batch)
        debug(f"Inserted {len(batch)} rows into {table} (total {total})")
        batch.clear()

    debug(f"Done {table}: {total} rows copied.")


def main() -> None:
    parser = argparse.ArgumentParser(description="Copy tables from MSSQL to MySQL (read-only source)")
    parser.add_argument("--config", required=True, help="Path to JSON config")
    parser.add_argument("--tables", help="Comma-separated override list of tables to copy")
    args = parser.parse_args()

    cfg = load_config(args.config)
    source = cfg["source"]
    target = cfg["target"]
    options = cfg["options"]
    tables: List[str] = []
    if args.tables:
        tables = [t.strip() for t in args.tables.split(",") if t.strip()]
    else:
        tables = cfg.get("tables", [])
    if not tables:
        raise SystemExit("No tables specified. Use --tables or config.tables")

    mssql_cn = connect_mssql(source)
    mysql_cn = connect_mysql(target)
    try:
        for t in tables:
            copy_table(mssql_cn, mysql_cn, source["database"], source.get("schema", "dbo"), t, options)
    finally:
        try:
            mysql_cn.close()
        except Exception:
            pass
        try:
            mssql_cn.close()
        except Exception:
            pass


if __name__ == "__main__":
    main()

File: scripts/mssql_to_mysql_copy.php
<?php
/**
 * PHP CLI: Read-only MSSQL → MySQL copier for selected tables.
 *
 * Usage examples:
 *  php scripts/mssql_to_mysql_copy.php --config scripts/config.gemas_dys.local.json
 *  php scripts/mssql_to_mysql_copy.php --config scripts/config.gemas_dys.local.json --tables CariEtiketTanimi,StokOzellikleri --batch-size 2000 --overwrite 0
 *
 * Notes:
 * - Source is MSSQL via PDO sqlsrv; only SELECT is used (read-only behavior).
 * - Destination is MySQL via PDO mysql; tables created if missing using type mapping.
 * - Primary keys are detected from MSSQL INFORMATION_SCHEMA and applied on MySQL.
 */

ini_set('display_errors', '1');
error_reporting(E_ALL);

function abort(string $msg): void {
    fwrite(STDERR, $msg . "\n");
    exit(1);
}

function info(string $msg): void {
    fwrite(STDOUT, $msg . "\n");
}

function loadConfig(string $path): array {
    if (!is_file($path)) {
        abort("Config not found: {$path}");
    }
    $raw = file_get_contents($path);
    if ($raw === false) {
        abort("Cannot read config: {$path}");
    }
    $cfg = json_decode($raw, true);
    if (!is_array($cfg)) {
        abort("Invalid JSON in config: {$path}");
    }
    $cfg['source'] = $cfg['source'] ?? [];
    $cfg['target'] = $cfg['target'] ?? [];
    $cfg['options'] = $cfg['options'] ?? [];
    $cfg['tables'] = $cfg['tables'] ?? [];
    $cfg['source']['schema'] = $cfg['source']['schema'] ?? 'dbo';
    $cfg['options']['batch_size'] = (int)($cfg['options']['batch_size'] ?? 1000);
    $cfg['options']['overwrite'] = (bool)($cfg['options']['overwrite'] ?? false);
    $cfg['options']['engine'] = $cfg['options']['engine'] ?? 'InnoDB';
    $cfg['options']['charset'] = $cfg['options']['charset'] ?? 'utf8mb4';
    $cfg['options']['collation'] = $cfg['options']['collation'] ?? null;
    return $cfg;
}

function parseArgs(): array {
    $longopts = [
        'config:',
        'tables::',
        'batch-size::',
        'overwrite::',
    ];
    $args = getopt('', $longopts);
    if (!isset($args['config'])) {
        abort('Usage: --config <path> [--tables <comma,list>] [--batch-size <n>] [--overwrite 0|1]');
    }
    return $args;
}

function connectMssql(array $source): PDO {
    $host = $source['server'] ?? null;
    $port = $source['port'] ?? null;
    $db   = $source['database'] ?? null;
    $user = $source['user'] ?? null;
    $pass = $source['password'] ?? null;
    if (!$host || !$db || !$user || $pass === null) {
        abort('source.server, source.database, source.user, source.password are required');
    }
    // Accept both "host,port" or host + port fields
    if ($port && strpos($host, ',') === false) {
        $host = $host . ',' . $port;
    }
    $drivers = PDO::getAvailableDrivers();
    if (!in_array('sqlsrv', $drivers, true)) {
        $driversStr = $drivers ? implode(', ', $drivers) : '(none)';
        abort(
            "PDO SQLSRV driver not available. PDO drivers: {$driversStr}.\n" .
            "Enable the pdo_sqlsrv extension in the php.ini used by this CLI.\n" .
            "Tip: run 'php --ini' to see the loaded php.ini; 'php -m' should list pdo_sqlsrv and sqlsrv."
        );
    }
    $dsn = "sqlsrv:Server={$host};Database={$db}";
    $opts = [PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION];
    if (defined('PDO::SQLSRV_ATTR_ENCODING') && defined('PDO::SQLSRV_ENCODING_UTF8')) {
        $opts[PDO::SQLSRV_ATTR_ENCODING] = PDO::SQLSRV_ENCODING_UTF8;
    }
    $pdo = new PDO($dsn, $user, $pass, $opts);
    return $pdo;
}

function connectMysql(array $target): PDO {
    $host = $target['host'] ?? 'localhost';
    $port = (int)($target['port'] ?? 3306);
    $db   = $target['database'] ?? null;
    $user = $target['user'] ?? null;
    $pass = $target['password'] ?? null;
    $charset = $target['charset'] ?? 'utf8mb4';
    if (!$db || !$user || $pass === null) {
        abort('target.database, target.user, target.password are required');
    }
    $dsn = "mysql:host={$host};port={$port};dbname={$db};charset={$charset}";
    $opts = [
        PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
        PDO::ATTR_EMULATE_PREPARES => false,
        PDO::MYSQL_ATTR_INIT_COMMAND => "SET NAMES {$charset}",
    ];
    return new PDO($dsn, $user, $pass, $opts);
}

function qi(string $name): string { return '`' . str_replace('`', '``', $name) . '`'; }

function fetchColumnsMssql(PDO $mssql, string $db, string $schema, string $table): array {
    $sql = "
        SELECT COLUMN_NAME, DATA_TYPE, CHARACTER_MAXIMUM_LENGTH,
               NUMERIC_PRECISION, NUMERIC_SCALE, IS_NULLABLE
        FROM INFORMATION_SCHEMA.COLUMNS
        WHERE TABLE_CATALOG = :db AND TABLE_SCHEMA = :sch AND TABLE_NAME = :tbl
        ORDER BY ORDINAL_POSITION
    ";
    $st = $mssql->prepare($sql);
    $st->execute([':db'=>$db, ':sch'=>$schema, ':tbl'=>$table]);
    $rows = $st->fetchAll(PDO::FETCH_ASSOC);
    if (!$rows) {
        abort("No columns found for {$schema}.{$table} in {$db}");
    }
    return $rows;
}

function fetchPkMssql(PDO $mssql, string $db, string $schema, string $table): array {
    $sql = "
        SELECT kcu.COLUMN_NAME
        FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS tc
        JOIN INFORMATION_SCHEMA.KEY_COLUMN_USAGE kcu
          ON tc.CONSTRAINT_NAME = kcu.CONSTRAINT_NAME
         AND tc.TABLE_SCHEMA = kcu.TABLE_SCHEMA
         AND tc.TABLE_NAME = kcu.TABLE_NAME
        WHERE tc.TABLE_CATALOG = :db AND tc.TABLE_SCHEMA = :sch
          AND tc.TABLE_NAME = :tbl AND tc.CONSTRAINT_TYPE = 'PRIMARY KEY'
        ORDER BY kcu.ORDINAL_POSITION
    ";
    $st = $mssql->prepare($sql);
    $st->execute([':db'=>$db, ':sch'=>$schema, ':tbl'=>$table]);
    return array_map(fn($r) => $r['COLUMN_NAME'], $st->fetchAll(PDO::FETCH_ASSOC));
}

function mapType(array $col): string {
    $dt = strtolower($col['DATA_TYPE']);
    $len = $col['CHARACTER_MAXIMUM_LENGTH'];
    $prec = $col['NUMERIC_PRECISION'];
    $scale = $col['NUMERIC_SCALE'];

    if ($dt === 'int') return 'INT';
    if ($dt === 'bigint') return 'BIGINT';
    if ($dt === 'smallint') return 'SMALLINT';
    if ($dt === 'tinyint') return 'TINYINT';
    if ($dt === 'bit') return 'TINYINT(1)';
    if ($dt === 'real') return 'FLOAT';
    if ($dt === 'float') return 'DOUBLE';
    if (in_array($dt, ['decimal','numeric','money','smallmoney'], true)) {
        $p = $prec !== null ? (int)$prec : 19;
        $s = $scale !== null ? (int)$scale : 4;
        return "DECIMAL({$p},{$s})";
    }
    if ($dt === 'char' || $dt === 'nchar') {
        if ((int)$len === -1) return 'LONGTEXT';
        return 'CHAR(' . (int)$len . ')';
    }
    if ($dt === 'varchar' || $dt === 'nvarchar') {
        if ((int)$len === -1) return 'LONGTEXT';
        return 'VARCHAR(' . (int)$len . ')';
    }
    if ($dt === 'text' || $dt === 'ntext') return 'LONGTEXT';
    if ($dt === 'varbinary') {
        if ((int)$len === -1) return 'LONGBLOB';
        return 'VARBINARY(' . (int)$len . ')';
    }
    if ($dt === 'binary') return 'BINARY(' . ((int)$len > 0 ? (int)$len : 1) . ')';
    if ($dt === 'image') return 'LONGBLOB';
    if ($dt === 'date') return 'DATE';
    if ($dt === 'smalldatetime' || $dt === 'datetime') return 'DATETIME';
    if ($dt === 'datetime2') return 'DATETIME(6)';
    if ($dt === 'time') return 'TIME';
    if ($dt === 'uniqueidentifier') return 'CHAR(36)';
    if ($dt === 'xml') return 'LONGTEXT';
    return 'LONGTEXT';
}

function mysqlTableExists(PDO $mysql, string $table): bool {
    // MySQL native prepared statements cannot prepare SHOW statements; build SQL safely
    $sql = 'SHOW TABLES LIKE ' . $mysql->quote($table);
    $st = $mysql->query($sql);
    return $st && $st->fetchColumn() !== false;
}

function buildCreateTable(string $table, array $columns, array $pkCols, array $opts): string {
    $defs = [];
    foreach ($columns as $col) {
        $name = $col['COLUMN_NAME'];
        $nullable = strtoupper((string)$col['IS_NULLABLE']) === 'YES';
        $type = mapType($col);
        $defs[] = qi($name) . ' ' . $type . ' ' . ($nullable ? 'NULL' : 'NOT NULL');
    }
    $pk = '';
    if ($pkCols) {
        $pk = ', PRIMARY KEY (' . implode(', ', array_map('qi', $pkCols)) . ')';
    }
    $engine = $opts['engine'] ?? 'InnoDB';
    $charset = $opts['charset'] ?? 'utf8mb4';
    $collation = $opts['collation'] ?? null;
    $collateSql = $collation ? ' COLLATE=' . $collation : '';
    $sql = 'CREATE TABLE ' . qi($table) . " (\n  " . implode(",\n  ", $defs) . $pk .
           "\n) ENGINE={$engine} DEFAULT CHARSET={$charset}{$collateSql};";
    return $sql;
}

function copyTable(PDO $mssql, PDO $mysql, string $db, string $schema, string $table, array $opts): void {
    info("\n=== {$schema}.{$table} ===");
    $columns = fetchColumnsMssql($mssql, $db, $schema, $table);
    $pkCols = fetchPkMssql($mssql, $db, $schema, $table);

    $exists = mysqlTableExists($mysql, $table);
    if (!$exists) {
        $ddl = buildCreateTable($table, $columns, $pkCols, $opts);
        info("Creating table {$table} in MySQL...");
        $mysql->exec($ddl);
    } else {
        if (!empty($opts['overwrite'])) {
            info("Overwriting table {$table} (drop + recreate)...");
            $mysql->exec('DROP TABLE ' . qi($table));
            $ddl = buildCreateTable($table, $columns, $pkCols, $opts);
            $mysql->exec($ddl);
        } else {
            info("Table {$table} already exists; appending rows.");
        }
    }

    $colNames = array_map(fn($c) => $c['COLUMN_NAME'], $columns);
    $select = 'SELECT ' . implode(', ', array_map(fn($n) => '['.$n.']', $colNames)) . ' FROM ['.$schema.'].['.$table.']';
    $placeholders = implode(', ', array_fill(0, count($colNames), '?'));
    $insert = 'INSERT INTO ' . qi($table) . ' (' . implode(', ', array_map('qi', $colNames)) . ") VALUES (" . $placeholders . ')';

    $batchSize = (int)($opts['batch_size'] ?? 1000);
    $insertStmt = $mysql->prepare($insert);

    // Begin copying
    $total = 0;
    $mssql->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
    $mysql->beginTransaction();
    $selStmt = $mssql->prepare($select, [PDO::ATTR_CURSOR => PDO::CURSOR_FWDONLY]);
    $selStmt->execute();

    $buffer = [];
    while ($row = $selStmt->fetch(PDO::FETCH_NUM)) {
        $buffer[] = $row;
        if (count($buffer) >= $batchSize) {
            foreach ($buffer as $r) { $insertStmt->execute($r); }
            $mysql->commit();
            $total += count($buffer);
            info("Inserted " . count($buffer) . " rows into {$table} (total {$total})");
            $mysql->beginTransaction();
            $buffer = [];
        }
    }
    if ($buffer) {
        foreach ($buffer as $r) { $insertStmt->execute($r); }
        $mysql->commit();
        $total += count($buffer);
        info("Inserted " . count($buffer) . " rows into {$table} (total {$total})");
    }
    info("Done {$table}: {$total} rows copied.");
}

// Main
$args = parseArgs();
$cfg = loadConfig($args['config']);
if (!empty($args['tables'])) {
    $cfg['tables'] = array_values(array_filter(array_map('trim', explode(',', $args['tables']))));
}
if (!empty($args['batch-size'])) { $cfg['options']['batch_size'] = (int)$args['batch-size']; }
if (isset($args['overwrite'])) { $cfg['options']['overwrite'] = (bool)((int)$args['overwrite']); }

$source = $cfg['source'];
$target = $cfg['target'];
$options = $cfg['options'];
$tables = $cfg['tables'];
if (!$tables) { abort('No tables specified (config.tables or --tables)'); }

info('Connecting MSSQL (read-only usage) ...');
$mssql = connectMssql($source);
info('Connecting MySQL (target) ...');
$mysql = connectMysql($target);

try {
    foreach ($tables as $t) {
        copyTable($mssql, $mysql, $source['database'], $source['schema'] ?? 'dbo', $t, $options);
    }
} finally {
    $mssql = null; // close
    $mysql = null;
}

info("All done.");
